### DISCUSSION TITLE
Not resuming events in Elsa 3.1.3

### DISCUSSION THREAD

**Discussion by zuntio** [elsa-workflows/elsa-core#5610] [Q&A]
Getting a bit desperate here.

So first of all I have implemented this kind of base composition which runs the actual activity and if not success, blocks execution with Event.

```csharp
public class StepComposition<TActivity, TOutput> : Composite<TOutput>
    where TActivity : AbstractActivity<TOutput>, new()
{
    /// <summary>
    /// EventName to distinct blocking events by 
    /// </summary>
    private Input<string> BlockingEventName { get; set; }

    private readonly Variable<ResultWrapper<TOutput>> _outputReference = new();

    public StepComposition() : this(0)
    {
    }

    public StepComposition(int order) : this(order, new())
    {
    }

    public StepComposition(int order, TActivity startingActivity)
    {
        // Setup metadata
        startingActivity = startingActivity.WithActivityMetadata<TOutput, TActivity>(order);
        Metadata = startingActivity.Metadata;

        // Setup input
        BlockingEventName = new(context => GetBlockingEventName(context, startingActivity));

        // Setup activities
        startingActivity.Result = new(_outputReference); // Save result in a variable
        var promptForManualRetryDecision = new FlowDecision(context =>
            startingActivity
                .GetResult<ResultWrapper<TOutput>>(context)?.WasSuccess ??
            false); // Force prompt if no correct output
        var waitForUserInputEvent = new Event(BlockingEventName);
        var manualRetryDecision = new FlowDecision(context =>
            waitForUserInputEvent.GetResult<RetryDecision>(context) is RetryDecision
                .Retry); // Retry only if explicit value found
        var fault = new Fault();
        var complete = new Complete();

        Variables =
        [
            _outputReference,
        ]; // OBS! Do not set variable references in FlowCharts scope. It will be done by base class

        // Arrange
        Root = new Flowchart
        {
            Start = startingActivity,
            Activities =
            {
                startingActivity,
                promptForManualRetryDecision,
                waitForUserInputEvent,
                manualRetryDecision,
                fault,
                complete,
            },
            Connections =
            {
                new(startingActivity, promptForManualRetryDecision), // From beginning to check if success
                new(new Endpoint(promptForManualRetryDecision, bool.FalseString),
                    new(waitForUserInputEvent)), // No success so prompt
                new(new Endpoint(promptForManualRetryDecision, bool.TrueString), new(complete)), // Success so complete
                new(new Endpoint(manualRetryDecision, bool.TrueString),
                    new(startingActivity)), // Decide to try again, start over
                new(new Endpoint(manualRetryDecision, bool.FalseString),
                    new(fault)), // Don't try again, straight to Fault status
            },
        };
    }

    private static string GetBlockingEventName(ExpressionExecutionContext context, TActivity activity)
    {
        var wfiName = context.GetWorkflowExecutionContext().GetProperty<string>(SetName.WorkflowInstanceNameKey);
        var stepNamePart = activity.GetEventNamePart();

        var sb = new StringBuilder();

        if (string.IsNullOrWhiteSpace(wfiName))
            sb.Append($"UNKNOWN-AT-{DateTime.UtcNow.Ticks}");
        else
            sb.Append(wfiName);
        sb.Append('-');
        sb.Append(stepNamePart);

        return sb.ToString();
    }

    protected override void OnCompleted(ActivityCompletedContext context)
    {
        context.WorkflowExecutionContext.AddExecutionLogEntry("Executiontimes fetched!");
        var result = _outputReference.Get<ResultWrapper<TOutput>>(context.TargetContext);
        context.TargetContext.Set(Result, result is not null ? result.Result : default);
    }
}
```

If event is found in separate query, the decision is then done in the following endpoint:
```csharp
public class Post : ElsaEndpoint<Request>
{
    private readonly IEventPublisher _eventPublisher;

    public Post(IEventPublisher eventPublisher)
    {
        _eventPublisher = eventPublisher;
    }

    /// <inheritdoc />
    public override void Configure()
    {
        Post(PostFlowdecision.Path);
        ConfigurePermissions(SpecialPermissions);
    }

    /// <inheritdoc />
    public override async Task HandleAsync(Request req, CancellationToken ct)
    {
        if (req.EventName is not { Length: > 0 })
            AddError(r => r.EventName, "Event name should be non-empty string");

        if (!Enum.IsDefined(req.Decision))
            AddError(r => r.Decision, $"Decision must be in range of enum {nameof(EventBookmarksCommon.RetryDecision)}");

        if (ValidationFailed)
        {
            await SendErrorsAsync(cancellation: ct);
            return;
        }

        await _eventPublisher.DispatchAsync(eventName: req.EventName!,
            workflowInstanceId: req.WorkflowInstanceId,
            activityInstanceId: req.ActivityInstanceId,
            payload: req.Decision,
            cancellationToken: ct);

        await SendOkAsync(ct);
    }
}
```

Request (enum also used in first snippet):
```csharp
    public class Request
    {
        public EventBookmarksCommon.RetryDecision Decision { get; set; }
        public string? EventName { get; set; }
        public string? ActivityInstanceId { get; set; }
        public string? WorkflowInstanceId { get; set; }
    }
    public enum RetryDecision
    {
        Retry = 1,
        Cancel = 2,
    }
```

But when I do this, all that happens is that row appears to `WorkflowInboxMessages` table and log shows `DispatchWorkflowDefinitionCommand completed with no result`.

What am I doing wrong here? Temporal stuff works just fine and nothing special is defined at `elsa.UseWorkflowManagement` and `elsa.UseWorkflowRuntime` invocations. Only telling there to use EF and SqlServer.

**Comment by glime-ai** [comment]
Based on the code snippets and the behavior you're describing, it seems like you're trying to implement a custom composite activity (`StepComposition<TActivity, TOutput>`) that executes an activity and, upon failure, blocks the workflow execution until an external event is received to either retry or cancel the operation.

The issue you're encountering, where a row appears in the `WorkflowInboxMessages` table and the log shows `DispatchWorkflowDefinitionCommand completed with no result`, suggests that the event dispatching mechanism is not triggering the continuation of the workflow as expected.

Here are a few things to check and consider:

1. **Event Name Matching**: Ensure that the event name you're dispatching (`req.EventName`) matches the event name expected by the `Event` activity within your composite activity. The event name is used to correlate the incoming event with the correct workflow instance and bookmark.

2. **Workflow Correlation**: Verify that the workflow instance ID and activity instance ID you're passing with the event (`req.WorkflowInstanceId` and `req.ActivityInstanceId`) are correct. These IDs are crucial for the workflow engine to identify which workflow and activity to resume.

3. **Event Dispatching**: Make sure that the event dispatching is done correctly. The `_eventPublisher.DispatchAsync` method should be properly implemented to trigger the event that the workflow is waiting for. If there's an issue with the event dispatching, the workflow will not resume.

4. **Workflow Configuration**: Double-check your workflow configuration to ensure that the `Event` activity is set up correctly to listen for the event and that the workflow is properly persisted and can be resumed.

5. **Error Handling**: Implement error handling in your endpoint to catch any exceptions that may occur during the event dispatching process. This can help identify if there's an issue with the event dispatching logic.

6. **Elsa Version**: Make sure you're using a compatible version of Elsa with your implementation. If you're using features from a newer version of Elsa, ensure that your project references that version.

7. **Debugging**: Use debugging tools to step through the code and monitor the state of the workflow. Check the state of the workflow before and after the event is dispatched to see if it's being resumed as expected.

If after checking these points the issue persists, you may want to provide more context or seek assistance from the Elsa community or support channels.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).

**Reply by zuntio** [reply]
I am positive that every check here is correct. To be sure, can you provide me the bare minimum of the Program.cs to make Event dispatching to work?

**Comment by zuntio** [comment]
There were missing connection in workflow definition and also needed to go to different direction. Closing.
